### ISSUE TITLE
Enhancement: support tilemap creation from Tiled

### ISSUE THREAD

**Issue by CJLove** [yeastplume/aloevera#20]
It would be great to have the option of creating a tilemap in aloevera directly from a tilemap defined via [Tiled](https://www.mapeditor.org/).  Right now this can be done by exporting a png from Tiled and importing the png image as an imageset in aloevera. There are a couple of ways this could be done:

1. Parse the Tiled `.tmx` file directly.  The file format is documented here: https://doc.mapeditor.org/en/stable/reference/tmx-map-format/.  A basic crate exists: https://docs.rs/tiled/0.9.2/tiled/
2. Parse the tilemap's json representation exported from Tiled.  This format is documented here: https://doc.mapeditor.org/en/stable/reference/json-map-format/.

Both of these approaches allow aloevera to leverage Tiled's support for specifying horizontally and vertically flipped tiles.  Option #1 has the advantage of eliminating the manual "export" step from Tiled.

I'm not sure how different this tilemap import processing would be from what happens now in the `tilemap init` and `tilemap load` steps.  Any thoughts?

**Comment by yeastplume** [comment]
Yes, absolutely this would be a great thing to have. I was using [Tilesetter](https://store.steampowered.com/app/1105890/Tilesetter/) to auto-generate basic tiles and maps for the sample data, but Tiled is an obvious target to support for import.

So quick thoughts:

* Would definitely parse the `.tmx` file directly, would be more convenient for everyone and a crate exists that seems to have been updated fairly recently.

* To ensure other formats can be introduced for import later, I'd make sure the import process looked and worked exactly like the existing PNG import from a code perspective. The way Aloevera is constructed is to maintain an internal representation of data that maps as cleanly as possible to what VERA expects, then have importers that create that internal representation and assemblers that transform the data into final formats. So I think the cleanest approach would be to run the `palette import` function on the `.tmx` file, then run the `imageset create/format`, `tilemap create` etc. (So there'd be an `load_from_tmx` function for each in addition to the existing`load_from_png`).  Not only will this ensure that everything is consistent for future import formats, it will also mean no additional error handling and format checking will need to be defined since it's already there. Of course, a new helper command can be added to the cli to run these all at once on a `.tmx` file, but internally they'd be following the same workflow as a png imports, 

* I'd probably want to factor out the `load/import` functions into Traits to define a convenient interface for new file type imports.

* Rotations need to be handled by the internal structures first, and I'd meant to get around to this at some stage. It should be a relatively simple addition. Right now on imageset import/format, a hash is stored for each frame. When the tilemap is imported, the hash of each tile in the map is checked against each frame in the imageset and the matching frame number is stored. So all that needs to happen here is that after the imageset data is import and formatted, a hash should be stored for the other 3 possible rotations of each tile along with which rotation the tile represents. Then the tile import can check all rotation hashes of each frame for a match, and the formatters can output the rotation values.

* With all sorts of different import methods being added, it'll become even more important to start filling out the integration tests as development ticks along. I've added a framework in `tests` that simulates invoking the command-line internally and checks the results. So whenever some functionality is added or changed, the output should be inspected manually (loaded into the emulator, checked however) then the correct output should be turned into a test by adding the reference data to the tests and checking. Might be a bit painful to maintain as changes happen, but it will help to ensure nothing gets unintentionally broken.

* Also, the samples are currently broken against the current revision of the emulator, might as well fix these or it'll be harder to do manual checks of reference data output.

So a quick series of steps (this even feels like a roadmap):

* Fix enough of the samples to allow for manual testing and inspection of results against the current emulator
* Add support for tile rotation while filling out enough imageset and tilemap integration tests to exercise it
* Design common Trait(s) for import functions and refactor all import functions into them. Palettes can already be imported from two separate types of file, so would also take this opportunity to create automated tests that ensure the output from both types of palette import is identical.
* Add support for `.tmx` palette import, then imageset import/format, tileset etc until done. At this stage it should be easy to create automated tests that check the results against existing reference output.

So.. seems like quite a few steps but I think following this approach would get us support for tile rotation, better integration tests, a coherent manner of adding new file formats, and direct support for Tiled `.tmx` files while setting the project up to handle future import formats. Sounds like a good short-term roadmap.

If you're interested in handling any of these steps, let me know and we can coordinate, (and no time pressures for any of this, this is all hobby coding on a long background thread)

**Comment by CJLove** [comment]
I'll help where I can, though some of this goes beyond my current quarantine-fueled rust skills :).  For starters I can probably get the samples updated for r37, as I've done this for my own code and the cc65 sprite sample in the `x16-demo` repo.

From an (automated) testing standpoint one thing to explore at least on Linux is to run the emulator headless under `xrdb-run` (X virtual frame buffer) and with the `-dump V` command-line argument to dump VRAM.  The x16 test program that runs can trigger an emulator exit by doing `jmp $ffff` on completion and then the script running the tests can then validate VRAM contents against expected values.  I do something like this (albeit no VRAM validation) in this [runTest.py](https://github.com/CJLove/libX16Test/blob/master/bin/runTest.py) python script in my libX16Test unit testing library.

**Comment by yeastplume** [comment]
> I'll help where I can, though some of this goes beyond my current quarantine-fueled rust skills :). For starters I can probably get the samples updated for r37, as I've done this for my own code and the cc65 sprite sample in the `x16-demo` repo.

That is a brilliant start, and I can see you've already had a go at that, will look at it now and merge. Will also take a look at the internal tile rotation and trait refactoring when I have some spare time, hopefully this week.
 
> From an (automated) testing standpoint one thing to explore at least on Linux is to run the emulator headless under `xrdb-run` (X virtual frame buffer) and with the `-dump V` command-line argument to dump VRAM. The x16 test program that runs can trigger an emulator exit by doing `jmp $ffff` on completion and then the script running the tests can then validate VRAM contents against expected values. I do something like this (albeit no VRAM validation) in this [runTest.py](https://github.com/CJLove/libX16Test/blob/master/bin/runTest.py) python script in my libX16Test unit testing library.

Very good idea, and I'd thought along similar lines of checking emulator output. A lot of work to do to get there, and might be slightly less frustrating to wait until a more final version of the emulator to start doing this, but I think this is definitely an approach we should be taking long-term.
